### ISSUE TITLE
recipeSet issues on materials branch

### ISSUE THREAD

**Issue by DavidBrainard** [ISET/iset3d#33]
This may be known and understood.  The function recipeSet has many options.  For some, it returns the empty matrix.  This is because the first thing it does is set its return variable, out, equal to [].  Then, for many of its options, it operates on its input recipe (thisR), but never sets the result to out.

It may be this is because receipeSet is on its way out the door as part of the materials branch.  I encountered it because I was going to update t_piIntro_simpletransforms to work on the materials branch (seemed useful).  But this call early on breaks:

thisR = recipeSet(thisR,'outputfile',...
    fullfile(piRootPath,'local','coloredCube','coloredCube.pbrt'));

Since it may be there is already a plan on how to handle this, I decided to leave it alone.

**Comment by digital-pro** [comment]
David – Great timing. I just tripped over the same thing this morning and was wondering what was up with the two very different return strategies. I was going to ask Brian later today. Drove me nuts for a little while until I figured out the issue. 

 

In the larger scheme of things now that Matlab supports code to implement properties, in future I’m not sure we need set/get methods. I guess maybe those were left over from before Recipes were objects?

**Comment by DavidBrainard** [comment]
This was enough of a clue to get me to try
thisR.set('outputfile', fullfile(piRootPath,'local','coloredCube','coloredCube.pbrt'));
which appears to work.

Probably recipeSet is being deprecated and a fix like this is appropriate in places where
it appears.

**Comment by DavidBrainard** [comment]
I got further in my efforts to revive t_piIntro_simpletransforms, but code that appears
to work in tls_assets.mlx does not seem to do so in the simpletransforms script.  For
me this crashes out when I try to set the rotation on the cube, line 48 of the script.

It may be that the coloredCube recipe is not set up right, although it reads in and
I can print the tree structure out just fine.

Maybe I'm not generalizing from the tls_assets script correctly, or maybe this
is revealing some buglet that wasn't tripped when a similar thing was done with
the simple scene recipe.

Best,

David

**Comment by digital-pro** [comment]
David – Try this:

thisR.set('asset', thisAsset.name, 'rotate', [0, 0, 45]);

 

 

*	David
*	]

**Comment by DavidBrainard** [comment]
Great, works!

What led me astray was that I overgeneralized from the translate example in tls_assets (pasted below),
and assumed that rotate could take an asset itself.

In general, I realize I'm not clear on what may be passed for the asset, when.  And indeed, not clear on
when we apply transformations to an asset itself and when to the parent (again, as distinguished
in the translate and rotate examples in tls_assets.)

Best,

David

% This is the object representing the yellow man
assetName = 'figure_6m_O';

% In this example, we find the branch node that is just above the yellow
% man, representing its position, rotation and such.
thisAsset = thisR.get('asset parent id',assetName);

% We add a translation, moving yellow man 2 meters in the z direction.
translateBranch = thisR.set('asset', thisAsset, 'translate', [0, 0, -2]);

**Comment by digital-pro** [comment]
One thing that might be helpful in the scripts/tutorials is if we try to distinguish between things and their names/IDs (I’m probably as guilty as the next person, but maybe I can make it a New Year’s Resolution). For example the code David was looking at below includes:

 

thisAsset = thisR.get('asset parent id',assetName); 

% We add a translation, moving yellow man 2 meters in the z direction. 
translateBranch = thisR.set('asset', thisAsset, 'translate', [0, 0, -2]); 



Which looking at it I’d interpret as getting an object handle/instance back from the get call and passing it. Except it is actually just an ID for the asset in the tree. Does this make sense?

**Comment by DavidBrainard** [comment]
I've now reconstructed more of the path that led me astray.

I had started with the name, but I had also started trying to rotation what
I thought was the whole cube, name '003ID_Cube_B'.  Using this
as the asset name in simpletransforms still crashes, but I guess
for a reason other than that it is the name and not the asset.  That
led me to try passing the asset rather than the name ...

Using '005ID_Cube_O' runs but only rotates one side of the cube (because, well,
it is one side).

So, more mysteries.  Can one only rotate a leaf?

**Comment by DavidBrainard** [comment]
Found and fixed the underlying bug in piAssetRotate.  There was a vestigial call of form 
   thisR = thisR.set(...)
in the part of the code where a branch was being rotated.  This didn't show up in tests where a leaf was being rotated (as in tls_assets).

Also fixed the examples in piAssetRotate and piAssetTranslate so they run, and did a little comment cleaning.

Commit 9ef02ffa on the materials branch.

Closing this issue.

**Comment by wandell** [comment]
Good.  DJC and I are just dealing with this issue at a slightly more global scale. Will update when we think our solution is working correctly.

Brian

**Comment by digital-pro** [comment]
David – That’s great. While you were doing that, Brian and I were taking a look at recipeSet in general. There has been some understandable confusion (including yours and mine) over when it returns a recipe (which it probably doesn’t need to, since it is a call-by-reference class) and when it returns some type of other value. So we modified it to return a tuple of [thisR, out]. So callers expecting a recipe will get one. Callers expecting some other type of return need to be modified to [~, out]. Seems like the best short-term fix. We ran through the asset test script and fixed the one caller that broke. Holler if you find anything awful that we broke! – David
